src/misc/camera.ts

```diff
@@ -0,0 +1,156 @@
+    .filter(({ kind }) => kind === 'videoinput')
+    .filter(({ label }) => !deviceBlackList.includes(label))
+    .filter(({ label }) => !label.includes('infrared'))
+
+  if (devices.length > 2) {
+    // Explicitly picking the first entry in the list of all videoinput
+    // devices for as the default front camera and the last entry as the default
+    // rear camera seems to be a good heuristic on some devices.
+    const frontCamera = devices[0]
+    const rearCamera = devices[devices.length - 1]
+
+    switch (camera) {
+      case 'auto':
+        return { deviceId: { exact: rearCamera.deviceId } }
+      case 'rear':
+        return { deviceId: { exact: rearCamera.deviceId } }
+      case 'front':
+        return { deviceId: { exact: frontCamera.deviceId } }
+      default:
+        return undefined
+    }
+  } else {
+    switch (camera) {
+      case 'auto':
+        return { facingMode: { ideal: 'environment' } }
+      case 'rear':
+        return { facingMode: { exact: 'environment' } }
+      case 'front':
+        return { facingMode: { exact: 'user' } }
+      default:
+        return undefined
+    }
+  }
+}
+
+export default async function (
+  videoEl: Ref<HTMLVideoElement | undefined>,
+  { camera, torch }: { camera: string; torch: string }
+) {
+  if (!videoEl.value) return
+
+  // At least in Chrome `navigator.mediaDevices` is undefined when the page is
+  // loaded using HTTP rather than HTTPS. Thus `STREAM_API_NOT_SUPPORTED` is
+  // initialized with `false` although the API might actually be supported.
+  // So although `getUserMedia` already should have a built-in mechanism to
+  // detect insecure context (by throwing `NotAllowedError`), we have to do a
+  // manual check before even calling `getUserMedia`.
+  if (window.isSecureContext !== true) {
+    throw new InsecureContextError()
+  }
+
+  if (navigator?.mediaDevices?.getUserMedia === undefined) {
+    throw new StreamApiNotSupportedError()
+  }
+
+  // This is a browser API only shim. It patches the global window object which
+  // is not available during SSR. So we lazily apply this shim at runtime.
+  await shimGetUserMedia()
+
+  const constraints = {
+    audio: false,
+    video: {
+      width: { min: 360, ideal: 640, max: 1920 },
+      height: { min: 240, ideal: 480, max: 1080 },
+      ...(await narrowDownFacingMode(camera))
+    }
+  }
+
+  const stream = await navigator.mediaDevices.getUserMedia(constraints)
+
+  // if (videoEl.value.srcObject) {
+  videoEl.value.srcObject = stream
+  //   // @ts-ignore
+  // } else if (videoEl.value.mozSrcObject !== undefined) {
+  //   // @ts-ignore
+  //   videoEl.value.mozSrcObject = stream
+  // } else if (window.URL.createObjectURL) {
+  //   // @ts-ignore
+  //   videoEl.value.src = window.URL.createObjectURL(stream)
+  // } else if (window.webkitURL) {
+  //   // @ts-ignore
+  //   videoEl.value.src = window.webkitURL.createObjectURL(stream)
+  // } else {
+  //   videoEl.value.src = stream.id
+  // }
```

Comment:
Typescript wasn't happy about those statements. Some googling showed that `mozSrcObject` does not exist anymore, for example, so I removed it for now. Would need to try edge case browsers, but for a new major release (dropping Vue 2 support e.g.) that could be added later, I thought.

Reply:
Reverted those changes

Reply:
> mozSrcObject does not exist anymore

Right, but the idea is to support legacy browsers. 

> When I comment the disabled code back in, I get an error in Firefox and Chrome.

What kind of error?

Reply:
Chrome: `DOMException: Could not start video source`
Firefox: `TypeError: URL.createObjectURL: Argument 1 is not valid for any of the 1-argument overloads.` (at `camera.ts:127`)

Reply:
Hmm weird :thinking: can you reproduce this on the current demo page?

Reply:
Do you mean I should push the commented-in code to the demo page so you can check as well? The mentioned code is currently not part of the deployment.

Reply:
No I mean the master branch demo page:

https://gruhn.github.io/vue-qrcode-reader/

If the same issue occurs there, I guess it’s really a problem with this code in particular. Otherwise it must be some interaction with other changes. And that might be worrisome.

Reply:
I do not have this problem on the current demo page, you're right. Maybe it's related to the new bundling. I'll need to check, but am currently on vacation so it might take a fair bit of time until I get around to that :pray:

Reply:
Alright, other than that I’m done complaining 😅

Reply:
Nevermind, I think I found the issue. Before it was:
```
if (videoEl.srcObject !== undefined)
```
which is supposed to check whether the property is defined in the browser at all. If `srcObject` is defined but empty, it's value is `null`. You had
```
if (videoEl.srcObject)
```
and that's guaranteed to fail.